src/debug/startSession.ts

```diff
@@ -0,0 +1,111 @@
+import * as vscode from 'vscode'
+import getPort from 'get-port'
+import { log } from '../log'
+import { getConfig } from '../config'
+import type { GlobalTestRunner } from '../runner/runner'
+import type { VitestAPI } from '../api'
+
+export interface DebugSessionAPI {
+  session: vscode.DebugSession | undefined
+  stop: () => Promise<void>
+}
+
+const DEBUG_DEFAULT_PORT = 9229
+
+export async function startDebugSession(
+  api: VitestAPI,
+  runner: GlobalTestRunner,
+  request: vscode.TestRunRequest,
+  token: vscode.CancellationToken,
+): Promise<DebugSessionAPI> {
+  let mainSession: vscode.DebugSession | undefined
+  const onStartDispose = vscode.debug.onDidStartDebugSession((session) => {
+    mainSession = session
+    onStartDispose.dispose()
+  })
+  const onRerunDispose = vscode.debug.onDidStartDebugSession(async (session) => {
+    const baseSession = session.parentSession || session
+    // check if not terminated
+    setTimeout(() => {
+      if (baseSession === mainSession) {
+        runner.runTests(request, token)
+          .catch((e) => {
+            if (!e.message.includes('timeout on calling'))
+              log.error('Error while running tests', e)
+          })
+      }
+    }, 150)
+  })
+  const onTerminateDispose = vscode.debug.onDidTerminateDebugSession(async (session) => {
+    if (mainSession !== session)
+      return
+
+    let timeout = false
+    let restarted = false
+    const onNewStartDispose = vscode.debug.onDidStartDebugSession((session) => {
+      onNewStartDispose.dispose()
+      // session.
+      if (timeout)
+        return
+
+      restarted = true
+      mainSession = session
+    })
+
+    setTimeout(() => {
+      if (!restarted) {
+        timeout = true
+        onTerminateDispose.dispose()
+        onNewStartDispose.dispose()
```

Comment:
These calls, are they guaranteed to not throw exceptions? Because if not, I feel like we should wrap them in try-catches individually, so we always clean everything up even if one of them fails.

Reply:
They are guaranteed